scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java

```diff
@@ -225,21 +301,63 @@ public synchronized boolean put(final Connection connection) {
 	}
 
 	@Override
-	public synchronized boolean update(final Connection connection) {
-		if (connection != null && connections.update(connection.getPeerAddress())) {
+	public synchronized boolean update(final Connection connection, InetSocketAddress newPeerAddress) {
+		if (connection == null) {
+			return false;
+		}
+		if (connections.update(connection.getConnectionId())) {
+			if (!connection.equalsPeerAddress(newPeerAddress)) {
+				InetSocketAddress oldPeerAddress = connection.getPeerAddress();
+				LOG.debug("{}connection: {} updated, address changed from {} to {}!", tag, connection.getConnectionId(),
+						oldPeerAddress, newPeerAddress);
+				if (oldPeerAddress != null) {
+					connectionsByAddress.remove(oldPeerAddress, connection);
+				}
+				connection.setPeerAddress(null);
```

Comment:
FMPOV, the logic of an address change is:
- remove the old address (that's the set(null) ).
- add the new address (that's the  set(new))

If the address of a connection is "only" removed, that must also cancel the flight.

In my opinion, that's straight forward.

Reply:
I will insist a bit.

Intuitively your code and codes below sound to be the same :
```java
connection.setPeerAddress(newPeerAddress);
if (newPeerAddress != null) {
     addToAddressConnections(connection);
}
```
or 
```java 
if (newPeerAddress != null) {
     connection.setPeerAddress(newPeerAddress);
     addToAddressConnections(connection);
} else {
     connection.setPeerAddress(null);
}
```

But as I understand this is not the same because the setter is not just a setter and we need to do a `connection.setPeerAddress(null)` before `connection.setPeerAddress(newPeerAddress)`, right ?

Reply:
The point will be:
you can't change the address of the connection with `setPeerAddress` without changing it in the connection store. It MUST be done by the connection store `update`.

Reply:
And yes, with that it's not longer a simple setter.

Reply:
Maybe it's more about mention that in the javadoc.

Reply:
Maybe, in a future PR, we remove the cancel flight from the connection at all.

There sooo many good changes ahead, yesterday I found the reassembling of the fragments is required to be improved. So, I'm not sure, if we should spend now too much time in this. 
Usually, if I try out a different design/approach, it's not only about running the unit test once, I run them on three different fast machines several hundred times and it takes a couple of days to ensure, that it's not worse than before. So I would prefer to shift this discussion and merge it with the new javadoc.

Reply:
> It MUST be done by the connection store update.
> Maybe it's more about mention that in the javadoc.

:+1: 

Just a last question about that, why `connection.setPeerAddress(newPeerAddress)`
does not cancel Flight and handshake when `newPeerAddress` != null ?

Reply:
According the current RFC for the connection ID, the address could only be changed **after** the Handshake. So, one pain point is the looong discussion within the IETF, the other is the amount of change lines for this redesign. I feel not too comfortable to have all this changes open.

Reply:
> According the current RFC for the connection ID, the address could only be changed after the Handshake. 

So that would make sense to always cancel flight / handshake when address changes.

Anyway let's this aside. It's not a big deal.

Reply:
When the address is changed, a session is established and so no flight is left to be cancelled.
Anyway, the cancel will be removed from the connection. In the past there may have been reasons to cancel the flight ASAP. In the meantime, with the serialized execution, it could be reduced to the handshaker's next flight or termination. But that requires time to do, and even more time to test. Therefore, I would prefer to shift it into the future.